Battery voltage throttle compensation

bnmoore opened the pull request:
Battery voltage throttle compensation.  Increases the throttle slightly in response to battery voltage decreases during flight.  Uses code similar to vbat pid compensation.  Intended for 1S micro quads.

etracer65 commented on src/main/sensors/battery.c:
```diff
@@ -426,6 +426,14 @@ float calculateVbatPidCompensation(void) {
+float calculateVbatThrottleCompensation(void) {
+    float batteryScaler =  1.0f;
+    if (batteryConfig()->voltageMeterSource != VOLTAGE_METER_NONE && batteryCellCount > 0) {
+        batteryScaler =  constrainf((( (float)batteryConfig()->vbatmaxcellvoltage * batteryCellCount ) / (float) voltageMeter.filtered), 1.0f, 1.6f);
```
I'm not sure about the hardcoded 1.6x max boost.  Perhaps this should be tuneable as a configuration parameter.  160% is quite a maximum boost and may not be suitable to all applications.  Also the only way to tune this currently would be to adjust the minimum cell voltage but this would also affect the vbat PID comp.

ledvinap replied:
Also using max cell voltage means that compensation will be always larger that one. Maybe using reference voltage for 100% throttle would work better

etracer65 commented:
Also check the unit test failure involving conflicting data types.

https://travis-ci.org/betaflight/betaflight/jobs/357707178

etracer65 commented:
This looks like a good idea and with a little cleanup and improvement it could be very useful to brushed micro flyers as well as possibly usable by racers looking to manage battery sag towards the end of heats.

ledvinap commented on src/main/flight/mixer.c:
```diff
@@ -721,7 +721,7 @@ static void applyMixToMotors(float motorMix[MAX_SUPPORTED_MOTORS])
 }
 
-void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation)
+void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation, uint8_t vbatThrottleCompensation)
```
`bool vbatThrottleCompensation` ?

ledvinap commented on src/main/flight/pid.h:
```diff
@@ -82,6 +82,7 @@ typedef struct pidProfile_s {
     uint16_t pidSumLimit;
     uint16_t pidSumLimitYaw;
     uint8_t vbatPidCompensation;            // Scale PIDsum to battery voltage
+    uint8_t vbatThrottleCompensation;       // Scale throttle to battery voltage
```
This breaks PG. New value should do to end of struct (or version has to be incremented)

mikeller commented:
I like this idea too. I think, as suggested, this should be changed to use `vbat_throttle_gain_percent` and `vbat_throttle_gain_voltage` (as the starting voltage for the compensation).

bnmoore commented:
Thank you for the suggestions, I've tried to make most of these changes.  I opted for parameters to control the "start" or "ideal" voltage and the max allowed throttle compensation.  I also changed the logic so that when voltage is high, the throttle is reduced.  I am assuming the motor rpm is linear with voltage (motor kv) and there is a voltage where flight characteristics are optimal.  Initial values are now gain 0 (off) and voltage 40 (4.0 volts).

mikeller commented:
@bnmoore: I am not sure that your latest change is what @etracer65 suggested. Instead of reducing the throttle above the starting voltage, it should just be linear, so that the throttle boost has a 'knee' when the configured start voltage is reached, like so:

![image](https://user-images.githubusercontent.com/4742747/37871888-c36c6744-3055-11e8-9dc0-a97c634e6afc.png)

joelucid commented:
Seems like a potential battery killer to me. Might be safer if a minimum voltage after sag could be configured and throttle wouldn’t be boosted beyond that. In particular with brushed which often don’t have telemetry.

bnmoore commented:
@mikeller @etracer65  Understood.  In the code I can change the low constraint on the throttle compensation to achieve that, but which way is preferred?  Throttle at high voltage may be too jumpy for some.  I am okay with either option.

![chart](https://user-images.githubusercontent.com/32078082/37871941-38c3aa30-2fb0-11e8-9a39-e4caf997b84a.png)

mikeller commented:
@joelucid: Good point. I think this should at least take the low voltage cutoff (https://github.com/betaflight/betaflight/blob/9edce2359a62cd56fa9787114df84868b7436843/src/main/fc/fc_rc.c#L320-L322), and stop boosting when that one kicks in.

bnmoore commented:
@mikeller Does the existing low voltage cutoff protect against this? or do I need to add a check to the throttle compensation code?

mikeller commented:
@bnmoore: It does, but only if it is enabled. The default is to reduce the throttle to 100 percent when BATTERY_CRITICAL is reached (i.e. no reduction).

etracer65 commented:
I've been reviewing the comments and coding so far and have a few suggestions to help simplify and get this moving forward.

I feel like we need 4 adjustable parameters:

1. `vbat_throttle_gain` (ON|OFF) - enables/disables the feature.  I realize this is a reversion and we could use the gain = 0 to disable, but I'll explain why later.
2. `vbat_throttle_gain_percent` (1 - 100)
3.  `vbat_throttle_gain_max_voltage` (33 - 43) - Cell voltage in 1/10 volt steps where throttle gain starts
4.  `vbat_throttle_gain_min_voltage` (33 - 43) - Cell voltage in 1/10 volt steps where throttle gain stops

I envision the logic as follows:

- Voltage above `vbat_throttle_gain_max_voltage` - gain is 1.0
- Voltage between `vbat_throttle_gain_max_voltage` and `vbat_throttle_gain_min_voltage` - gain is `1.0 + ((vbat_throttle_gain_max_voltage - current_voltage) / (vbat_throttle_gain_max_voltage - vbat_throttle_gain_min_voltage)) * (vbat_throttle_gain_percent / 100)`
- Voltage below `vbat_throttle_gain_min_voltage` - gain is `1.0 + (vbat_throttle_gain_percent / 100)`

So assuming:
`vbat_throttle_gain_percent = 50` (50% max boost)
`vbat_throttle_gain_max_voltage = 40` (4.0 volts)
`vbat_throttle_gain_min_voltage = 35` (3.5 volts)

This gives a throttle boost curve like:

![throttle_boost_curve](https://user-images.githubusercontent.com/17088539/37982236-96a3a9a8-31df-11e8-99b9-b524a19bf96d.png)


To address earlier concerns...

Yes, this could potentially be a "battery killer".  But I think we need to leave that up to the pilot to decide if it makes sense for them to use.  If the throttle boost didn't hold at the low volt cutoff, the I think it would make it very difficult on the pilot as voltage sag temporarily drops below the limit playing havoc with their throttle control.  In an effort to save the battery, using the low voltage cutoff logic with its time-based hysteresis should be fine.  I'm not sure we need the additional application of LVC in the throttle boost calculation and should be able simply use the existing logic in fc_rc.c as it already scales the throttle input which in turn will scale the boosted throttle.  But this could be discussed further.

Finally, the rationale for reverting to using a `vbat_throttle_gain = ON|OFF` control rather than basing on `vbat_throttle_gain_percent > 0` is that I feel this type of logic has the potential to be extended to `vbat_pid_gain`.  One of the common complaints I hear from racer pilots is that they don't use `vbat_pid_gain` because it's not adjustable enough and they can't get a tune that avoids oscillations when the battery is nearly full.  It has a fixed gain of 1.33 and begins scaling directly from max cell voltage - both things we wanted to avoid with throttle boost.  So while improvements to `vbat_pid_gain` should be a separate PR, having throttle boost structured as described will allow consistency between both features.

ledvinap commented:
@etracer65 : Having minimum gain less than 100% can be practical too - 
make throttle range identical even on full battery (limiting throttle)
(maybe) protect motors overload on full battery

etracer65 commented:
@ledvinap While I see your point, I start to wonder where we reach the crossover between usability and too much complexity.  For example, I'm wondering how a pilot would go about tuning in a scenario where we had minimum gain less than 1.0 and maximum gain.  I'm not against it, just wondering if the additional parameter and additional complexity is worth it vs. the practicality of actually using this capability.

ledvinap commented:
@etracer65 : Another possibility is only to place 100% throttle voltage.

Assuming ESC is using linear PWM based on input and neglecting some effects (current drop, commutation loss, etc.), RPMs are directly proportional to PWM and voltage (rpm = V<sub>act</sub>* pwm). Then PWM boost should be V<sub>ref</sub>/V<sub>act</sub> (as done in this PR)

Maybe no limits are needed - minimum boost is at full battery voltage, it is quite close to working voltage, so throttle reduction won't be that large (and power would be slightly larger than on V<sub>ref</sub> due to other effects). 
Maximum boost limit may be derived from warning voltage - you want to notify user that battery is running dry, that's exactly what warning voltage is for. Ramp to no throttle boost over half of V<sub>warn</sub> and V<sub>critical</sub> may work fine.

(BTW: boost graph is hyperbola, not line)

borisbstyle commented:
Changed the label to MAJOR FEATURE.

Remember that modifying throttle may jeoperdize safety. There are many cases where this can lead into fly aways or unpredictable behaviour.

Some impact analysis should be performed on:
Failsafe scenarios need to be considered, where throttle is preset
False battery inputs can lead into change of altitude
etc...

etracer65 commented:
To continue on the safety track, another concern would be the user choosing a voltage range and boost the might lead to making it difficult to land or even descend depending on the battery state.  At high gains the margin above minimum throttle might be so small that it's difficult to avoid climbing.  So we need to add in some mechanism to scale in the boost effect by throttle position.  Maybe a linear application of the "boost amount" when pilot throttle input is from 0-50% and full application above 50%.

mikeller commented:
Is this going to be a headline feature for 3.4? If not, it should not be labelled as a major feature...

ledvinap commented:
@etracer65 : Using V<sub>ref</sub>/V<sub>act</sub>, effective power will be very close to reference (voltage seen by motor is almost the same). So unless you choose extreme reference voltage (well over 4.2V/cell), it will only compensate battery voltage.

bnmoore commented:
I've simplified this substantially and went with 3 parameters.  
**vbatThrottleCompensationEnabled** turns the feature ON/OFF.  
**vbatThrottleCompensationVoltage** is the "ideal" voltage that we are trying to emulate. Setting this near Vmax causes all positive gain throughout the flight, vice versa, and everything in between.  I reverted to hard coded limits of 0.5 to 1.5 for max allowed throttle gain.  
**vbatThrottleCompensationSlope** is the slope for the voltage : throttle compensation (lower slope is less correction)  The default values (voltage = 33, slope = 60) work well for me on my 1S micro.

Re: additional safety checks.  I agree, there should be more, but as the conversation shows, this is not an easy task.  I won't have time to put that level of work into it.  IMO, one should not fly past the limits of their battery.

ledvinap commented on src/main/flight/mixer.c:
```diff
@@ -744,6 +744,14 @@ void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation)
+    if(currentPidProfile->vbatThrottleCompensationEnabled) {     
+        float slope = 200.0f / currentPidProfile->vbatThrottleCompensationSlope;
+        float vbatThrottleCompensationFactor = 
+            constrainf( (currentPidProfile->vbatThrottleCompensationVoltage - getBatteryAverageCellVoltage()) / slope, 0.5f, 1.5f );
```
Physically correct compensation is V<sub>ref</sub>/V<sub>act</sub>

It's not far from line (http://www.wolframalpha.com/input/?i=3.9%2Fx+for+x+from+3.3+to+4.2), so linear approximation may work, but is a bit questionable.
Note that practical range ~ +=20%(over full V<sub>ref</sub> range), so compensation is relatively small

V<sub>ref</sub>/V<sub>act</sub> is approximated when slope = V<sub>ref</sub> (slope at V<sub>ref</sub> is correct).

This is a bit awkward to setup correctly using vbatThrottleCompensationSlope ... using scaling of V<sub>ref</sub>/V<sub>act</sub> may be better choice.

BTW: double division on slope is used unnecessarily ...

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -134,7 +134,10 @@ void resetPidProfile(pidProfile_t *pidProfile)
+        .iterm_rotation = false,
+        .vbatThrottleCompensationEnabled = 0,
+        .vbatThrottleCompensationSlope = 60, // Lower means less compensation
+        .vbatThrottleCompensationVoltage = 33 // voltage in 0.1V steps
```
Nominal battery voltage is 3.7V

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -134,7 +134,10 @@ void resetPidProfile(pidProfile_t *pidProfile)
-        .iterm_rotation = false                 
+        .iterm_rotation = false,
+        .vbatThrottleCompensationEnabled = 0,
+        .vbatThrottleCompensationSlope = 60, // Lower means less compensation
```
from above (2000/vbatThrottleCompensationVoltage)

mikeller commented:
OMNIBUS is overflowing, we need to find something to cull.

Rakinroul commented:
I only wanna to note my tip, didnt read whole history, my aplogogises.

I can imagine this feature will work in goal to keep same rpm on throttle , even when battery voltage coming down ?

Between full battery and empty battery is difference almost 20 percents, so therefore can be linear increase how battery will droping down, linear increasement on throttle from (in 4s case) 16.8 to 14V starting on 0 and ending on these almost 20 percents.
Then you will maintain same rpm on in example 60 percents throttle.

mikeller commented:
@bnmoore: Can you rebase on current master please? This should fix the flash size overflow.

bnmoore commented:
@mikeller rebased

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -732,6 +732,13 @@ void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation)
     const float vbatCompensationFactor = vbatPidCompensation ? calculateVbatPidCompensation() : 1.0f;
+    
+    // Apply battery voltage compensation to throttle
+    if(currentPidProfile->vbatThrottleCompensationEnabled) {     
```
Space after `if`.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -732,6 +732,13 @@ void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation)
+    // Apply battery voltage compensation to throttle
+    if(currentPidProfile->vbatThrottleCompensationEnabled) {     
+        float vbatThrottleCompensationFactor = 
+            constrainf( (float)currentPidProfile->vbatThrottleCompensationVoltage / getBatteryAverageCellVoltage(), 0.5f, 1.5f );
```
No spaces on the inside of braces.

etracer65 replied:
I think all we need to change is the low end of the constraint range to be 1.0f insetad of 0.5f.  That still allows a start point but eliminates the reduction above the threshold.

mikeller commented on src/main/flight/pid.c:
```diff
@@ -134,7 +134,9 @@ void resetPidProfile(pidProfile_t *pidProfile)
         .throttle_boost_cutoff = 15,
-        .iterm_rotation = false                 
+        .iterm_rotation = false,
+        .vbatThrottleCompensationEnabled = 0,
```
Use `false` here.

etracer65 commented on src/main/flight/pid.c:
```diff
@@ -134,7 +134,9 @@ void resetPidProfile(pidProfile_t *pidProfile)
-        .iterm_rotation = false                 
+        .iterm_rotation = false,
+        .vbatThrottleCompensationEnabled = 0,
+        .vbatThrottleCompensationVoltage = 35 // Target voltage in 0.1V steps
```
I'm not sure this is a good default value.  3.5v is the lower limit that you'd normally fly your batteries so if a pilot enables the feature and doesn't change the default they'll get a significant throttle reduction in the current implementation or no boost in the requested change to not perform reductions.  Maybe a default of 3.8v might be better.  This would give a give some improvement later in the battery life but not lead to crazy gains when initially turned on.

etracer65 commented on src/main/interface/settings.c:
```diff
@@ -729,6 +729,9 @@ const clivalue_t valueTable[] = {
     { "d_yaw",                      VAR_UINT8  | PROFILE_VALUE, .config.minmax = { 0, 200 }, PG_PID_PROFILE, offsetof(pidProfile_t, pid[PID_YAW].D) },
 
+    { "vbat_throttle_compensation_enabled", VAR_UINT8  | PROFILE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_PID_PROFILE, offsetof(pidProfile_t, vbatThrottleCompensationEnabled) },
+    { "vbat_throttle_compensation_voltage", VAR_UINT8 | PROFILE_VALUE, .config.minmax = { 20, 50 }, PG_PID_PROFILE, offsetof(pidProfile_t, vbatThrottleCompensationVoltage) },
```
The range for battery voltage should be more reasonable.  Having such a low bottom of 2.0v doesn't really make sense in the scenario that we're only boosting.  Maybe 3.2v would be a better minimum.  I guess having a high max value is okay as it lets the user boost a lot if they want - but it'll probably be initially confusing as people will wonder how their batteries could ever be at 5.0v/cell.

robertlacroix replied:
Is it worth combining the two settings? i.e. if voltage is > 0 it's enabled, otherwise disabled?

etracer65 commented on src/main/flight/mixer.c:
```diff
@@ -732,6 +732,13 @@ void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation)
+    
+    // Apply battery voltage compensation to throttle
+    if(currentPidProfile->vbatThrottleCompensationEnabled) {     
+        float vbatThrottleCompensationFactor = 
```
Should be `const`

mikeller commented:
@etracer65: Is this still valuable? If so, do you want to take this over?

etracer65 commented:
@mikeller I think for this to be viable it needs a new direction. I don't think the current logic is going to work well for one simple reason - sag. I feel like this will result in a positive feedback loop where the pilot increases throttle, battery sags, then the boost detects that and increases throttle, more sag, more boost, etc.

I think we would need to look at battery voltage over a longer averaging period to get a truer measurement of the battery state.  We don't want to react to transients.

I can try and work on this to see if it's viable, but I would likely start with a fresh PR.

mikeller commented:
@etracer65: Good point about the sag and negative feedback. I'll remove the milestone from this ticket but leave it open to track new input on this topic.

quiljw01 commented:
@junwoo091400  and another version..... everyone should combine efforts on this topic and make it happen!
